### ISSUE TITLE
cal_gradient_penalty for WGAN-GP not used?

### ISSUE THREAD

**Issue by zhulingchen** [junyanz/pytorch-CycleGAN-and-pix2pix#1163]
Hello,

I have a question regarding your implementation of WGAN-GP based pix2pix.

It seems you just used the WGAN loss
```
        elif self.gan_mode == 'wgangp':
            if target_is_real:
                loss = -prediction.mean()
            else:
                loss = prediction.mean()
```

but the gradient penalty term has not been used, i.e., the `cal_gradient_penalty` function in networks.py has never been called for pix2pix implementation.

Would you please let me know why?

Thank you very much!

**Comment by zhulingchen** [comment]
I found GP term has not used for CycleGAN either.

**Comment by junyanz** [comment]
It was not used in CycleGAN/pix2pix. I added it in case other people want to use it for their own development.

**Comment by zhulingchen** [comment]
> 
> 
> It was not used in CycleGAN/pix2pix. I added it in case other people want to use it for their own development.

Thank you @junyanz for the answer. I implemented the WGANGP in the following way. In my Pix2pixGAN class, I have the gradient penalty term calculation private method:

```
    def __get_gradient_penalty_loss(self, real, fake, constant=1.0):
        batch_size = real.shape[0]
        alpha = torch.rand(batch_size, 1, 1, 1)
        alpha = alpha.expand_as(real).to(self.device)
        interpolated = alpha * real + (1 - alpha) * fake
        interpolated.requires_grad_(True)
        dummy = torch.empty(batch_size, 0, self.config['image_rows'], self.config['image_cols']).to(self.device)  # to fit the discriminator input argument list
        disc_interpolated = self.discriminator(interpolated, dummy)
        grad_interpolated = torch.autograd.grad(outputs=disc_interpolated, inputs=interpolated,
                                                grad_outputs = torch.ones_like(disc_interpolated),
                                                create_graph = True, retain_graph = True, only_inputs = True)[0]
        grad_interpolated = grad_interpolated.view(batch_size, -1)  # flat the data
        grad_norm = torch.sqrt(torch.sum(grad_interpolated ** 2, dim=1) + 1e-16)
        return torch.mean((grad_norm - constant) ** 2)
```

Then, I calculate the gradient penalty term `loss_gp` in the following way:
```
loss_gp = self.config['lambda_gp'] * self.__get_gradient_penalty_loss(real=torch.cat([real_src, real_tgt], dim=1),
                                                                                          fake=torch.cat([real_src, fake_tgt.detach()], dim=1))
```

The point is, I think in pix2pix implementation using WGAN-GP loss with the "mixed" setting, the "real data" should be `torch.cat([real_src, real_tgt], dim=1)` and the "fake data" should be `torch.cat([real_src, fake_tgt.detach()], dim=1)`.

Do you think I am doing the correct way?

Thanks!

**Comment by zhulingchen** [comment]
Meanwhile, do you think WGAN-GP loss for pix2pix or CycleGAN would yield better visual performances than the vanilla GAN loss?

**Comment by junyanz** [comment]
We haven't observed the performance gain from our experiments.

**Comment by zhulingchen** [comment]
> 
> 
> We haven't observed the performance gain from our experiments.

Thank you very much! So how about LSGAN loss? Is it better than the vanilla GAN loss?

**Comment by junyanz** [comment]
LSGAN works better for CycleGAN. No big difference for pix2pix.
